src/DynamicResponsive.tsx

```diff
@@ -30,147 +35,157 @@ export interface ResponsiveProviderState {
+const setupMatchers = (mediaQueries: MediaQueries): MediaQueryMatchers => {
+  return Object.keys(mediaQueries).reduce(
+    (matchers, key) => ({
+      ...matchers,
+      [key]: window.matchMedia(mediaQueries[key]),
+    }),
+    {}
+  )
+}
+
+/**
+ * Uses the matchers to build a map of the states of each media query
+ */
+const checkMatchers = (
+  mediaQueryMatchers: MediaQueryMatchers
+): MediaQueryMatches => {
+  return Object.keys(mediaQueryMatchers).reduce(
+    (matches, key) => ({
+      ...matches,
+      [key]: mediaQueryMatchers[key].matches,
+    }),
+    {}
+  )
+}
+
+const isSupportedEnvironment = () => {
+  return (
+    typeof window !== "undefined" && typeof window.matchMedia !== "undefined"
+  )
+}
 
-/** TODO */
 export function createResponsiveComponents<M extends string>() {
   const ResponsiveContext = React.createContext({})
   ResponsiveContext.displayName = "Media.DynamicContext"
 
   const ResponsiveConsumer: React.FunctionComponent<
     React.ConsumerProps<MediaQueryMatches<M>>
   > = ResponsiveContext.Consumer as React.FunctionComponent<
-    React.ConsumerProps<any>
+    React.ConsumerProps<any> & { children: any }
   >
 
-  return {
-    Consumer: ResponsiveConsumer,
-    Provider: class ResponsiveProvider extends React.Component<
-      ResponsiveProviderProps<M>,
-      ResponsiveProviderState
-    > {
-      constructor(props: ResponsiveProviderProps<M>) {
-        super(props)
-        let mediaQueryMatchers: MediaQueryMatchers | undefined = undefined
-        let mediaQueryMatches: MediaQueryMatches
-
-        if (this.isSupportedEnvironment()) {
-          mediaQueryMatchers = this.setupMatchers(props.mediaQueries)
-          mediaQueryMatches = this.checkMatchers(mediaQueryMatchers)
-        } else {
-          mediaQueryMatches = Object.keys(props.mediaQueries).reduce(
-            (matches, key) => ({
-              ...matches,
-              [key]:
-                !!props.initialMatchingMediaQueries &&
-                props.initialMatchingMediaQueries.includes(key as M),
-            }),
-            {}
-          )
-        }
-
-        this.state = {
-          mediaQueryMatchers,
-          mediaQueryMatches,
-        }
-      }
-
-      isSupportedEnvironment = () => {
-        return (
-          typeof window !== "undefined" &&
-          typeof window.matchMedia !== "undefined"
-        )
-      }
-
-      /**
-       * Create an array of media matchers that can validate each media query
-       */
-      setupMatchers = (mediaQueries: MediaQueries): MediaQueryMatchers => {
-        return Object.keys(mediaQueries).reduce(
-          (matchers, key) => ({
-            ...matchers,
-            [key]: window.matchMedia(mediaQueries[key]),
-          }),
-          {}
-        )
+  const ResponsiveProvider: FC<ResponsiveProviderProps<M>> = ({
+    mediaQueries,
+    initialMatchingMediaQueries,
+    children,
+  }) => {
+    /**
+     * TODO: The class component lifecycle shouldComponentUpdate
```

Comment:
Since there's no clear way to check state within `React.memo` just going to skip this for now.